### PR TITLE
AEA-941 multiplexer/agent/connections refactoring

### PR DESCRIPTION
**Pull request by solarw**
## Proposed changes

multiplexer/agent/connections refactoring

## Types of changes

- [x] Bugfix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Checklist

- [x] I have read the [CONTRIBUTING](../CONTRIBUTING.md) doc
- [x] I am making a pull request against the `develop` branch (left side). Also you should start your branch off our `develop`.
- [x] Lint and unit tests pass locally with my changes
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have checked that code coverage does not decrease.
- [x] I have checked that the documentation about the `aea cli` tool works
- [x] I have added necessary documentation (if appropriate)
- [x] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

File: tests/test_aea.py [hunk 1]
```diff
@@ -484,8 +484,9 @@ def test_add_behaviour_dynamically():
     resources = Resources()
     identity = Identity(agent_name, address=wallet.addresses[DEFAULT_LEDGER])
     connection = _make_local_connection(identity.address, LocalNode())
-    agent = AEA(identity, wallet, resources, default_connection=connection.public_id,)
     resources.add_connection(connection)
```
Comment (DavidMinarsch) [inline tests/test_aea.py:487]: So we now enforce that `resources` is complete at time of construction of AEA.

File: aea/multiplexer.py [hunk 2]
```diff
@@ -95,11 +97,20 @@ def __init__(
...
         self._id_to_connection: Dict[PublicId, Connection] = {}
         self._default_connection: Optional[Connection] = None
-        self._initialize_connections_if_any(connections, default_connection_index)
 
+        if not default_connection and connections:
+            enforce(
+                len(connections) - 1 >= default_connection_index,
+                "default_connection_index os out of connections range!",
+            )
+            default_connection = connections[default_connection_index].connection_id
```
Comment (DavidMinarsch) [inline aea/multiplexer.py:111]: if we do provide a `default_connection` we should still check it is present in `connection`

File: aea/agent.py [hunk 3]
```diff
@@ -47,6 +46,10 @@ class Agent(AbstractAgent, WithLogger):
     }
     DEFAULT_RUNTIME: str = "threaded"
 
+    _runtime: BaseRuntime
```
Comment (DavidMinarsch) [inline aea/agent.py:49]: So the ABC is not aware of the concept of Runtime..

File: tests/test_aea.py (hunk 1 above)
Comment (solarw) [reply tests/test_aea.py:487]: @DavidMinarsch  at least connections

Comment (solarw) [reply tests/test_aea.py:487]: I will add a proper check  on agent init

File: aea/agent.py (hunk 3 above)
Comment (solarw) [reply aea/agent.py:49]: it's just type hint for Agent class, cause runtime now defined in dedicated method, not init

File: tests/test_aea.py (hunk 1 above)
Comment (solarw) [reply tests/test_aea.py:487]: done

**Review by DavidMinarsch — APPROVED** [review]
LGTM

**Review by marcofavorito — APPROVED** [review]
LGTM

**Pull request merged**